onnxruntime/core/common/path.h

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <vector>
+
+#include "core/common/common.h"
+#include "core/common/path_string.h"
+
+namespace onnxruntime {
+
+// Note: We should use the std::filesystem library after upgrading to C++17.
+
+/** A filesystem path. */
+class Path {
+ public:
+  Path() = default;
+  Path(const Path&) = default;
+  Path& operator=(const Path&) = default;
+  Path(Path&&) = default;
+  Path& operator=(Path&&) = default;
+
+  /** Parses a path from `path_str`. */
+  static Status Parse(const PathString& path_str, Path& path);
+  /** Parses a path from `path_str`. Throws on failure. */
+  static Path Parse(const PathString& path_str);
+
+  /** Gets a string representation of the path. */
+  PathString ToPathString() const;
+  /** Gets a string representation of the path's root path, if any. */
+  PathString GetRootPathString() const;
+  /** Gets the path components following the path root. */
+  const std::vector<PathString>& GetComponents() const { return components_; }
+
+  /** Whether the path is empty. */
+  bool IsEmpty() const;
+
+  /** Whether the path is absolute (refers unambiguously to a file location). */
+  bool IsAbsolute() const;
+  /** Whether the path is relative (not absolute). */
+  bool IsRelative() const { return !IsAbsolute(); }
+
+  /** Returns a copy of the path without the last component. */
+  Path ParentPath() const;
+
+  /**
+   * Normalizes the path.
+   * A normalized path is one with "."'s and ".."'s resolved.
+   * Note: This is a pure path computation with no filesystem access.
+   */
+  Path& Normalize();
+  /** Returns a normalized copy of the path. */
+  Path Normalized() const {
+    Path p{*this};
```

Comment:
nit: Normalize and Normalized are very similar. Maybe NormalizedPath would be more consistent with things like ParentPath